README.md
# package-bot
> Review new package PR on [melpa](https://github.com/melpa/melpa) automatically.

## Dependencies

Here are list of required softwares and npm packages.

### Softwares

* [NodeJS](https://nodejs.org/en/) & [npm](https://www.npmjs.com/products/teams?utm_source=adwords&utm_medium=ppc&utm_campaign=npmTeams2019Q2&utm_content=site&gclid=Cj0KCQjw6sHzBRCbARIsAF8FMpXNGnP2OczbVGv4lG7bh9sJ53BXLP5S3-JixP0rL3L9qZNWbR6r5k8aAjNaEALw_wcB)
* [Git](https://git-scm.com/downloads)
* [Emacs](https://www.gnu.org/software/emacs/download.html)
* [cURL](https://curl.haxx.se/)

### npm packages

* [axios](https://www.npmjs.com/package/axios)
* [forever](https://www.npmjs.com/package/forever)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.


### 2020-03-17

* Added full control over `header`/`footer` using Emacs Lisp.

### 2020-03-11

* First version release.

### 2020-03-10

* Added the template files and designs.


data.js
/**
 * $File: data.js $
 * $Date: 2020-03-09 18:18:38 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */

"use strict";

const fs = require('fs');

const config = require('./config');
const helper = require('./helper');


/* Save data structure. */
var status = {
  pr_id : -1,
  timestamp: ''
};


/**
 * Save status to data file.
 *
 * Keep the status so next time you bootup the service it wouldn't
 * forget what was the state last time we closed.
 */
function saveStatus() {
  status.timestamp = helper.getTimestamp();
  return new Promise((resolve, reject) => {
    fs.writeFile(config.DATA_PATH, JSON.stringify(status), function (err) {
      if (err)
        console.log('[ERROR] Error while saving data %s', err);
      else
        console.log("[INFO] Saved status file once => '%s'", config.DATA_PATH);;
      resolve();
    });
  });
}

/**
 * Load status from storage.
 *
 * Read the status data to revive our data to current service. The next
 * PR check will be the latest PR. Hence, we wouldn't make package review
 * twice.
 */
function loadStatus() {
  return new Promise((resolve, reject) => {
    fs.readFile(config.DATA_PATH, function (err, rw_data) {
      if (err)
        console.log('[ERROR] Error while loading status %s', err);
      else {
        console.log('[INFO] Reading status...');
        /* Put down the data. */
        {
          let data = JSON.parse(rw_data);
          status.pr_id = data.pr_id;
        }
      }
      resolve();
    });
  });
}

//------------------- Module Exports -------------------//

module.exports.status = status;

module.exports.loadStatus = loadStatus;
module.exports.saveStatus = saveStatus;


util.js
/**
 * $File: util.js $
 * $Date: 2020-03-09 23:44:53 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */

"use strict";

const { exec } = require("child_process");


/**
 * Clone a project from URL to PATH.
 * @param { string } url : URL of the project.
 * @param { string } path : Path to clone.
 */
function cloneProject(url, path) {
  let cmd = 'git clone ' + url +  ' ' + path;
  return new Promise((resolve, reject) => {
    exec(cmd, (error, stdout, stderr) => {
      if (error) {
        console.log('[ERROR] Clone project error: %s', error.message);
        resolve();
      }
      if (stderr) {
        console.log('[ERROR] Clone project error: %s', stderr);
        resolve();
      }
      console.log('[INFO] Cloning project from %s to %s,\n\n %s', url, path, stdout);
      resolve();
    });
  });
}

/**
 * Check if the command valid.
 * @param { string } cmd : Command to check installed.
 * @param { string } name : Name of the software to be prompt.
 */
function ensureCommand(cmd, name) {
  return new Promise((resolve, reject) => {
    exec(cmd, (error, stdout, stderr) => {
      if (error) {
        console.log('[ERROR] This program require %s to be installed %s', name, error.message);
        resolve();
      }
      if (stderr) {
        console.log('[ERROR] This program require %s to be installed %s', name, stderr);
        resolve();
      }
      console.log('[INFO] %s version checked.\n\n%s', name, stdout);
      resolve();
    });
  });
}

/**
 * Get timestamp.
 */
function getTimestamp() {
  return new Date().toISOString().replace(/T/, ' ').replace(/\..+/, '');
}

//------------------- Module Exports -------------------//

module.exports.cloneProject = cloneProject;
module.exports.ensureCommand = ensureCommand;
module.exports.getTimestamp = getTimestamp;


index.js
/**
 * $File: index.js $
 * $Date: 2020-03-07 23:55:25 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */

"use strict";

const fs = require('fs');
const axios = require('axios');
const { exec } = require("child_process");

const config = require('./config');

// Record down the highest PR's ID use to check if there are any new PR.
var highestPRId = -1;


/**
 * Ensure Emacs is installed.
 */
function ensureEmacs() {
  return new Promise((resolve, reject) => {
    exec('emacs --version', (error, stdout, stderr) => {
      if (error) {
        console.log('[ERROR] This program require Emacs to be installed %s', error.message);
        resolve();
      }
      if (stderr) {
        console.log('[ERROR] This program require Emacs to be installed %s', stderr);
        resolve();
      }
      console.log('[INFO] Emacs version checked.\n\n%s', stdout);
      resolve();
    });
  });
}

/**
 * Save status to data file.
 */
function saveStatus() {
  let data = {
    pr_id: highestPRId,
    timestamp: new Date().toISOString(),
  };
  return new Promise((resolve, reject) => {
    fs.writeFile(config.DATA_PATH, JSON.stringify(data), function (err) {
      if (err)
        console.log('[ERROR] Error while saving data %s', err);
      else
        console.log("[INFO] Saved status file once => '%s'", config.DATA_PATH);;
      resolve();
    });
  });
}

/**
 * Update the current status use to check if any packae needed to be check.
 */
function updateStatus() {
  axios.get('https://api.github.com/repos/melpa/melpa/pulls')
    .then(response => {
      console.log(response.data);
    })
    .catch(error => {
      console.log(error);
    });
}

/**
 * Main loop to check any PR need to be review.
 */
function doCheckPR() {

}

/**
 * Register any events you want here.
 */
function registerEvents() {
  return new Promise((resolve, reject) => {
    console.log('[INFO] Register event action');
    {
      setInterval(doCheckPR, config.CHECK_TIME);
      console.log('  - Register check PR event');
    }
    resolve();
  });
}

/**
 * Program entry point.
 */
function main() {
  return new Promise((resolve, reject) => {
    console.log("Start initializing ::: (%s)", config.APP_NAME);
    resolve();
  }).then(() => {  /* Ensure Emacs installed. */
    return ensureEmacs();
  }).then(() => {  /* Prepare data file. */
    return saveStatus();
  }).then(() => {  /* Register events. */
    return registerEvents();
  }).then(() => {
    console.log("Done loading :::");
  }).catch((err) => {
    console.log("Initializing fail :::", err);
  });
}
main();


config.js
/**
 * $File: config.js $
 * $Date: 2020-03-09 00:25:07 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */

"use strict";

const DEBUG           = true;

const APP_NAME        = 'package-bot - MELPA';

// Interval time to do PR review.
const CHECK_TIME      = 1000 * 60 * 30;

const DATA_PATH       = './status.dat';
const REVIEW_PATH     = './review/';            // Clone the review package to this path.
const REVIEW_SCRIPT   = './review.el';
const REQUEST_DATA    = './request.json';       // Temporary file to output the body form.

const OUTPUT_HEADER   = './output/header.txt';  // File that contains review information.
const OUTPUT_BODY     = './output/body.txt';    // File that contains review information.
const OUTPUT_FOOTER   = './output/footer.txt';  // File that contains review information.

const TEMPLATE_HEADER = './templates/header.md';
const TEMPLATE_BODY   = './templates/body.md';
const TEMPLATE_FOOTER = './templates/footer.md';

const USERNAME        = '';
const ACCESS_TOKEN    = '';

//------------------- Module Exports -------------------//

module.exports.DEBUG = DEBUG;
module.exports.CHECK_TIME = CHECK_TIME;
module.exports.APP_NAME = APP_NAME;
module.exports.DATA_PATH = DATA_PATH;
module.exports.REVIEW_PATH = REVIEW_PATH;
module.exports.REVIEW_SCRIPT = REVIEW_SCRIPT;
module.exports.REQUEST_DATA = REQUEST_DATA;
module.exports.OUTPUT_HEADER = OUTPUT_HEADER;
module.exports.OUTPUT_BODY = OUTPUT_BODY;
module.exports.OUTPUT_FOOTER = OUTPUT_FOOTER;
module.exports.TEMPLATE_HEADER = TEMPLATE_HEADER;
module.exports.TEMPLATE_BODY = TEMPLATE_BODY;
module.exports.TEMPLATE_FOOTER = TEMPLATE_FOOTER;
module.exports.USERNAME = USERNAME;
module.exports.ACCESS_TOKEN = ACCESS_TOKEN;


helper.js
/**
 * $File: helper.js $
 * $Date: 2020-03-09 23:44:53 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */

"use strict";

const fs = require('fs');
const path = require("path");
const childProcecss = require("child_process");


/**
 * Clone a project from URL to PATH.
 * @param { string } url : URL of the project.
 * @param { string } path : Path to clone.
 */
function cloneProject(url, path) {
  let cmd = 'git clone ' + url +  ' ' + path;
  childProcecss.execSync(cmd, (error, stdout, stderr) => {
    console.log('[INFO] Cloning project from %s to %s %s', url, path, stdout);
  });
}

/**
 * Check if the command valid.
 * @param { string } cmd : Command to check installed.
 * @param { string } name : Name of the software to be prompt.
 */
function ensureCommand(cmd, name) {
  return new Promise((resolve, reject) => {
    childProcecss.exec(cmd, (error, stdout, stderr) => {
      if (error) {
        console.log('[ERROR] This program require %s to be installed %s', name, error.message);
        resolve();
      }
      if (stderr) {
        console.log('[ERROR] This program require %s to be installed %s', name, stderr);
        resolve();
      }
      console.log('[INFO] %s version checked.\n\n%s', name, stdout);
      resolve();
    });
  });
}

/**
 * Get timestamp.
 */
function getTimestamp() {
  return new Date().toISOString().replace(/T/, ' ').replace(/\..+/, '');
}

/**
 * Remove directory.
 * @param { string } dirPath : Directory path.
 */
function removeDir(dirPath) {
  if (!fs.existsSync(dirPath)) return;

  let list = fs.readdirSync(dirPath);
  for (var index = 0; index < list.length; ++index) {
    let filename = path.join(dirPath, list[index]);
    let stat = fs.statSync(filename);

    if (filename == "." || filename == "..") {
      // do nothing for current and parent dir
    } else if (stat.isDirectory())
      removeDir(filename);
    else
      fs.unlinkSync(filename);
  }
  fs.rmdirSync(dirPath);
};

//------------------- Module Exports -------------------//

module.exports.cloneProject = cloneProject;
module.exports.ensureCommand = ensureCommand;
module.exports.getTimestamp = getTimestamp;
module.exports.removeDir = removeDir;


review.el

(require 'package)

(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))

;; To avoid initializing twice
(setq package-enable-at-startup nil)

;; Disable check signature while installing packages.
(setq package-check-signature nil)

;; initialize package.el
(package-initialize)

(require 'f)

(defun review-read-file (path)
  "Read a file from PATH."
  (with-temp-buffer
    (insert-file-contents path)
    (buffer-string)))

(defun review-write (str path &optional append)
  "Write STR to PATH file.
APPEND to the file."
  (write-region str nil path append))

(defun review-get-buffer (buf-name default-str)
  "Get the buffer string from BUF-NAME, if not found return DEFAULT-STR instead."
  (let ((buf-str (ignore-errors (with-current-buffer buf-name (buffer-string)))))
    (if buf-str buf-str default-str)))

(defun review-compile-log ()
  "Return compile log as string for current buffer."
  (ignore-errors (byte-compile-file (buffer-name)))
  (review-get-buffer "*Compile-Log*" "No issues found."))

(defun review-checkdoc ()
  "Return checkdoc as string for current buffer."
  (ignore-errors (checkdoc-file (buffer-file-name)))
  (review-get-buffer "*Warnings*" "No issues found."))

(defun review-package-lint ()
  "Return package lint as string for current buffer."
  (ignore-errors (package-lint-current-buffer))
  (ignore-errors (with-current-buffer "*Package-Lint*" (buffer-string))))

(defun review-do ()
  "Do the review package action."
  (let* ((file-name (substring (buffer-file-name)
                               (length (cdr (project-current)))
                               (length (buffer-file-name))))
         (cmp-log (review-compile-log))
         (pkg-lnt (review-package-lint))
         (ckdoc (review-checkdoc))
         ;; NOTE: Order should corresponds to the template file.
         (review-text (format template-body-str
                              file-name
                              checkdoc-version ckdoc
                              emacs-version cmp-log
                              pkg-lnt)))
    (review-write review-text output-body t)))

;; Prepare variables from `main.js'.
(progn
  (setq output-header (expand-file-name output-header))
  (setq output-body (expand-file-name output-body))
  (setq output-footer (expand-file-name output-footer))
  (setq template-header (expand-file-name template-header))
  (setq template-body (expand-file-name template-body))
  (setq template-footer (expand-file-name template-footer))
  (setq project-dir (expand-file-name project-dir)))
;; Prepare new variables.
(progn
  (setq template-header-str (review-read-file template-header))
  (setq template-body-str (review-read-file template-body))
  (setq template-footer-str (review-read-file template-footer)))
;; Clean up.
(progn
  (review-write "" output-header)
  (review-write "" output-body)
  (review-write "" output-footer))

(message "> Checking project directory: %s" project-dir)

;; Writing output header.
(review-write template-header-str output-header)

;; Writing output body.
(let ((dirs (f-directories project-dir nil t))
      (files-el '()))
  (push project-dir dirs)
  (dolist (dir dirs) (add-to-list 'load-path (expand-file-name dir)))
  (dolist (dir dirs)
    (setq files-el (directory-files dir nil "\\.el$"))
    (dolist (file files-el)
      (setq file (concat dir (if (string-match-p "/$" dir) "" "/") file))
      (message "> Checking file: '%s'" file)
      (find-file file)
      (review-do))))

;; Writing output footer.
(review-write template-footer-str output-footer)

(message "Done review package")


main.js
/**
 * $File: main.js $
 * $Date: 2020-03-07 23:55:25 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */

"use strict";

const fs = require('fs');
const util = require('util');
const childProcecss = require("child_process");

const axios = require('axios');

const config = require('./config');
const data = require('./data');
const helper = require('./helper');


// Use to store the newest PR Id. To check if we need to make any
// package review.
var latest_pr_id = -1;

// Use to clean up afterward.
var infos = [];


/**
 * Update the current status use to check if any packae needed to be check.
 */
function updateStatus() {
  return new Promise((resolve, reject) => {
    axios.get('https://api.github.com/repos/melpa/melpa/pulls')
      .then(response => {
        latest_pr_id = response.data[0].number;
        resolve();
      })
      .catch(error => {
        console.log('[ERROR] Error while updating PR status: %s', error);
        resolve();
      });
  });
}

/**
 * * Parse MELPA template.
 *
 * ```md
 * ### Brief summary of what the package does
 * , and more...
 * ```
 * @param { string } body : String of the message.
 */
function parseTemplate(body) {
  let info = {
    summary : '',     /* No use */
    link: '',
    association: '',  /* No use */
    upstream: '',     /* No use */
    checklist: '',    /* No use */

    name: '',         /* Name of the repository. */
    clone_path: '',   /* Record down the clone path. */
  };

  let link_title = '### Direct link to the package repository';

  if (body.lastIndexOf(link_title) !== -1) {
    let lines = body.split('\r\n').filter(function (elm) {
      return elm != '';  // Remove empty string from array.
    });

    info.link = lines[lines.indexOf(link_title) + 1];

    let url_path = info.link.split('/');
    let pos_offset = (info.link.slice(-1) == '/') ? 2 : 1;
    info.name = url_path[url_path.length - pos_offset];
  }

  return info;
}

/**
 * Get PR information by NUM.
 * @param { number } num : PR id/number.
 */
function getPRInfo(num) {
  console.log('[INFO] Get PR %s information', num);
  console.log('---------------------------------------------------------');
  return new Promise((resolve, reject) => {
    axios.get('https://api.github.com/repos/melpa/melpa/pulls/' + num)
      .then(response => {
        let body = response.data.body;
        let info = parseTemplate(body);

        if (info.link == '') {
          console.log('[INFO] Detect PR %s is not the new package, no need to review', num);
          resolve();
        }

        infos.push(info);

        info.clone_path = config.REVIEW_PATH + info.name + '/';
        helper.removeDir(info.clone_path);                // Remove before clone.
        helper.cloneProject(info.link, info.clone_path);  // Clone it.

        /* Start review progress for this PR. */
        {
          console.log('[INFO] Starting the review progress..');
          let review_cmd = util.format
          ('emacs --batch --eval "(progn %s)" -l "%s"',
           util.format('%s %s %s %s %s %s %s',
                       util.format('(setq project-dir \\"%s\\")', info.clone_path),
                       util.format('(setq output-header \\"%s\\")', config.OUTPUT_HEADER),
                       util.format('(setq output-body \\"%s\\")', config.OUTPUT_BODY),
                       util.format('(setq output-footer \\"%s\\")', config.OUTPUT_FOOTER),
                       util.format('(setq template-header \\"%s\\")', config.TEMPLATE_HEADER),
                       util.format('(setq template-body \\"%s\\")', config.TEMPLATE_BODY),
                       util.format('(setq template-footer \\"%s\\")', config.TEMPLATE_FOOTER)),
           config.REVIEW_SCRIPT);

          childProcecss.execSync(review_cmd);
        }

        /* Make PR comment. */
        {
          console.log('[INFO] Making package review to the PR..');
          let pr_param = {
            body: '',
          };

          let header = fs.readFileSync(config.OUTPUT_HEADER, 'utf-8');
          let body = fs.readFileSync(config.OUTPUT_BODY, 'utf-8');
          let footer = fs.readFileSync(config.OUTPUT_FOOTER, 'utf-8');
          pr_param.body = header + body + footer;

          fs.writeFileSync(config.REQUEST_DATA, JSON.stringify(pr_param));

          let url = 'https://api.github.com/repos/melpa/melpa/issues/' + num + '/comments';

          if (config.DEBUG)
            url = 'https://api.github.com/repos/jcs090218/package-bot/issues/2/comments';

          let comment_cmd = util.format
          ('curl -u %s:%s --header "Content-Type: application/json" --request POST --data \"@%s\" \"%s\"',
           config.USERNAME, config.ACCESS_TOKEN, config.REQUEST_DATA, url);

          childProcecss.execSync(comment_cmd, (error, stdout, stderr) => { });
        }

        resolve();
      })
      .catch(error => {
        console.log('[ERROR] Error while getting PR %s info: %s', num, error);
        resolve();
      });
  });
}

/**
 * Start the review loop.
 */
function startReview() {
  return new Promise((resolve, reject) => {
    resolve();
  }).then(() => {
    ++data.status.pr_id;
    return getPRInfo(data.status.pr_id);
  }).then(() => {
    if (data.status.pr_id < latest_pr_id)
      return startReview();
    // PR ID will eventaully be the lastest PR ID.
    data.status.pr_id = latest_pr_id;
    return 0;
  });
}

/**
 * Make pacakge review if needed to.
 */
function makeReview() {
  return new Promise((resolve, reject) => {
    resolve();
  }).then(() => {
    if (latest_pr_id == data.status.pr_id) {
      console.log("[INFO] No new package PR need to review");
      return 0;
    }
    return startReview();
  });
}

/**
 * Clean up the review process.
 */
function cleanup() {
  return new Promise((resolve, reject) => {
    console.log('[INFO] Cleaning up the process');
    infos.forEach(function (info) {
      helper.removeDir(info.clone_path);
    });
    infos = [];  // clean up the array.
    resolve();
  });
}

/**
 * Main loop to check any PR need to be review.
 */
function doCheckPR() {
  return new Promise((resolve, reject) => {
    console.log("Start checking PR package :::");
    resolve();
  }).then(() => {  /* Update PR status. */
    return updateStatus();
  }).then(() => {  /* Do pacakge review. */
    return makeReview();
  }).then(() => {  /* Save last status once. */
    return data.saveStatus();
  }).then(() => {  /* Do clean up. */
    return cleanup();
  }).then(() => {
    console.log("[INFO] Done review package at time: %s", helper.getTimestamp());
  }).catch((err) => {
    console.log("[ERROR] Failed check PR while processing", err);
  });
}

/**
 * Register any events you want here.
 */
function registerEvents() {
  return new Promise((resolve, reject) => {
    console.log('[INFO] Register event action');
    {
      setInterval(doCheckPR, config.CHECK_TIME);
      console.log('  - [ ] Register check PR event');
    }
    resolve();
  });
}

/**
 * Program entry point.
 */
function main() {
  return new Promise((resolve, reject) => {
    console.log("Start initializing ::: (%s)", config.APP_NAME);
    resolve();
  }).then(() => {  /* Ensure Git installed. */
    return helper.ensureCommand('git --version', 'Git');
  }).then(() => {  /* Ensure Emacs installed. */
    return helper.ensureCommand('emacs --version', 'Emacs');
  }).then(() => {  /* Prepare data file. */
    if (fs.existsSync(config.DATA_PATH))
      return data.loadStatus();
    else
      return data.saveStatus();
  }).then(() => {  /* Register events. */
    return registerEvents();
  }).then(() => {
    console.log("\nDone loading :::\n");
    if (config.DEBUG) doCheckPR();
  }).catch((err) => {
    console.log("\nInitializing fail :::\n%s", err);
  });
}
main();
